### ISSUE TITLE
A curl "Connection reset by peer" is unretryable

### ISSUE THREAD

**Issue by fredericgboutin-yapla** [caseyamcl/guzzle_retry_middleware#41]
## Detailed description

See https://github.com/guzzle/guzzle/issues/2733#issuecomment-2396757118

Basically I would like to auto-retry a ["connection reset by peer"](https://stackoverflow.com/a/1434506/17193362) and because Guzzle returns a `$reason` typed as a `RequestException` it is simply [ignored by the retry middleware](https://github.com/caseyamcl/guzzle_retry_middleware/blob/319829956ad61be3f48fba8f79e789b92538e56f/src/GuzzleRetryMiddleware.php#L213), even if I want to use a custom callback.

I see no work around this.

## Context

I'm getting Sentry events called "GuzzleHttp\Exception\RequestException
cURL error 56: OpenSSL SSL_read: Connection reset by peer, errno 104 (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)"

The callstack points to this code - https://github.com/guzzle/guzzle/blob/d281ed313b989f213357e3be1a179f02196ac99b/src/Handler/CurlFactory.php#L276

Since I have unit tests, I added the `RequestException` use-case and I successfully reproduce the issue but it seems I cannot easily fix it.

A connection reset by peer seems to me that it should be retried.

## Possible implementation

I suppose we cannot simply handle all `RequestException`, so I would add a specific case in the `onRejected` callback, probably something like,

```php
protected function onRejected(RequestInterface $request, array $options): callable
    {
        return function (Throwable $reason) use ($request, $options): PromiseInterface {
            // If was bad response exception, test if we retry based on the response headers
            if ($reason instanceof BadResponseException) {
                if ($this->shouldRetryHttpResponse($options, $request, $reason->getResponse())) {
                    return $this->doRetry($request, $options, $reason->getResponse(), $reason);
                }
            // If this was a connection exception, test to see if we should retry based on connect timeout rules
            } elseif ($reason instanceof ConnectException) {
                // If was another type of exception, test if we should retry based on timeout rules
                if ($this->shouldRetryConnectException($options, $request)) {
                    return $this->doRetry($request, $options, null, $reason);
                }
            } elseif ($reason instanceof \GuzzleHttp\Exception\RequestException) {
               $context = $reason->getHandlerContext();

               // If was another type of exception, test if we should retry based on timeout rules
               if ($context['errno'] === 104 && $this->shouldRetryConnectException($options, $request)) {
                    return $this->doRetry($request, $options, null, $reason);
               }
            }

            // If made it here, then we have decided not to retry the request
            // Future-proofing this; remove when bumping minimum Guzzle version to 7.0
            if (class_exists('\GuzzleHttp\Promise\Create')) {
                return \GuzzleHttp\Promise\Create::rejectionFor($reason);
            } else {
                return rejection_for($reason); /* @phpstan-ignore-line */
            }
        };
    }
```

Doing this locally makes my unit-test pass.

**Comment by fredericgboutin-yapla** [comment]
## Workaround

I've created a pull request but I'm really not sure if it is going to be ever addressed, so here is a "local" workaround one can use in the mean time,

```php
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Promise\PromiseInterface;
use Psr\Http\Message\RequestInterface;
use Throwable;

/**
 * Custom GuzzleRetryMiddleware class to handle connection reset by peer errors
 * @see https://github.com/caseyamcl/guzzle_retry_middleware/issues/41
 */
class GuzzleRetryMiddleware extends \GuzzleRetry\GuzzleRetryMiddleware
{
	protected function onRejected(RequestInterface $request, array $options): callable
    {
		$parentCallBack = parent::onRejected($request, $options);

        return function (Throwable $reason) use ($request, $options, $parentCallBack): PromiseInterface {
            if ($reason instanceof RequestException) {
				$context = $reason->getHandlerContext();
                $errNo = $context['errno'] ?? null;

                // Normally, we would not retry on request exceptions, but because a connection reset by peer
                // could have occured, we will retry on this specific error
                if ($errNo == 104 && $this->shouldRetryConnectException($options, $request)) {
                    return $this->doRetry($request, $options, null, $reason);
                }
			}
			
			return $parentCallBack($reason);
        };
    }
}
```

**Comment by caseyamcl** [comment]
Pulled in PR; released v2.12.0
